onnxruntime/python/onnxruntime_pybind_state.cc

```diff
@@ -74,6 +76,89 @@ static Env& platform_env = Env::Default();
 #pragma warning(push)
 #endif
 
+using PyCallback = std::function<void(std::vector<py::object>, py::object user_data, std::string)>;
+
+struct AsyncResource {
+  std::vector<OrtValue> feeds;
+  std::vector<const OrtValue*> feeds_raw;
+
+  std::vector<std::string> feed_names;
+  std::vector<const char*> feed_names_raw;
+
+  std::vector<OrtValue*> fetches_raw;
+
+  std::vector<std::string> fetch_names;
+  std::vector<const char*> fetch_names_raw;
+
+  RunOptions default_run_option;
+  PyCallback callback;
+  py::object user_data;
+
+  void ReserveFeeds(size_t sz) {
+    feeds.reserve(sz);
+    feeds_raw.reserve(sz);
+    feed_names.reserve(sz);
+    feed_names_raw.reserve(sz);
+  }
+
+  void ReserveFetches(size_t sz) {
+    fetches_raw.reserve(sz);
+    fetch_names.reserve(sz);
+    fetch_names_raw.reserve(sz);
+  }
+};
+
+void AsyncCallback(void* user_data, OrtValue** outputs, size_t num_outputs, OrtStatusPtr ort_status) {
+  ORT_ENFORCE(user_data, "user data must not be NULL for callback in python");
+  PyGILState_STATE gstate{};
+  bool ensure_called = false;
+  {
+    std::unique_ptr<AsyncResource> async_resource{reinterpret_cast<AsyncResource*>(user_data)};
+
+    Ort::Status status(ort_status);
+    std::vector<py::object> rfetch;
+
+    // return on error
+    if (!status.IsOK()) {
+      async_resource->callback(rfetch, async_resource->user_data, status.GetErrorMessage());
+      return;
+    }
+
+    rfetch.reserve(num_outputs);
+
+    auto fill_fetches = [&]() {
+      size_t pos = 0;
+      for (size_t ith = 0; ith < num_outputs; ++ith) {
+        const auto& fet = *outputs[ith];
+        if (fet.IsAllocated()) {
+          if (fet.IsTensor()) {
+            rfetch.push_back(AddTensorAsPyObj(fet, nullptr, nullptr));
+          } else if (fet.IsSparseTensor()) {
+            rfetch.push_back(GetPyObjectFromSparseTensor(pos, fet, nullptr));
+          } else {
+            rfetch.push_back(AddNonTensorAsPyObj(fet, nullptr, nullptr));
+          }
+        } else {
+          rfetch.push_back(py::none());
+        }
+        ++pos;
+      }
+    };
+
+    if (PyGILState_Check()) {
+      fill_fetches();
+      async_resource->callback(rfetch, async_resource->user_data, "");
+    } else {
+      // py::gil_scoped_acquire acquire;
+      gstate = PyGILState_Ensure();
+      ensure_called = true;
+      fill_fetches();
+      async_resource->callback(rfetch, async_resource->user_data, "");
+    }
+  }
+  if (ensure_called) PyGILState_Release(gstate);
```

Comment:
AsyncResource should be destructed while the GIL is held since it contains pybind objects. Hence the introduction of a new scope block and the call to Release after the scope block.